Source/WTF/wtf/darwin/TypeCastsOSObject.h

```diff
@@ -0,0 +1,224 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#import <CoreFoundation/CFBase.h>
+#import <objc/runtime.h>
+#import <wtf/OSObjectPtr.h>
+#import <wtf/cf/TypeCastsCF.h>
+
+// To add support for a new OSObject type:
+// 1. Import header that defines OSObject type.
+// 2. Add type to an existing WTF_OS_OBJECT_*_TYPES(M) macro, or create a new one.
+//    a. If a new macro is created, add macro to WTF_OS_OBJECT_TYPES(M).
+//    b. If a new macro is created, create OSObjectTypeCastTraits declarations.
+
+#import <Network/Network.h>
+#import <dispatch/dispatch.h>
+
+#define WTF_OS_OBJECT_DISPATCH_TYPES(M) \
+    M(dispatch_group) \
+    M(dispatch_object) \
+    M(dispatch_queue) \
+    M(dispatch_queue_global) \
+    M(dispatch_source)
+
+WTF_EXTERN_C_BEGIN
+#define WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT(TypeName) \
+struct TypeName##_s;
+WTF_OS_OBJECT_DISPATCH_TYPES(WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT)
+#undef WTF_DECLARE_OS_OBJECT_DISPATCH_BASE_STRUCT
+WTF_EXTERN_C_END
+
+#define WTF_OS_OBJECT_NETWORK_TYPES(M) \
+    M(nw_endpoint) \
+    M(nw_resolution_report) \
+    M(nw_path)
+
+WTF_EXTERN_C_BEGIN
+#define WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT(TypeName) \
+struct TypeName;
+WTF_OS_OBJECT_NETWORK_TYPES(WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT)
+#undef WTF_DECLARE_OS_OBJECT_NETWORK_BASE_STRUCT
+WTF_EXTERN_C_END
+
+#define WTF_OS_OBJECT_TYPES(M) \
+    WTF_OS_OBJECT_DISPATCH_TYPES(M) \
+    WTF_OS_OBJECT_NETWORK_TYPES(M)
+
+// Because ARC enablement is a compile-time choice, and we compile this header
+// both ways, we need a separate copy of our code when ARC is enabled.
+#if __has_feature(objc_arc)
+#define dynamicOSObjectCast dynamicOSObjectCastARC
+#define osObjectCast osObjectCastARC
+#endif
```

Comment:
I'd be tempted to reverse this since ARC is the future:
```suggestion
// Because ARC enablement is a compile-time choice, and we compile this header
// both ways, we need a separate copy of our code when ARC is enabled.
#if !__has_feature(objc_arc)
#define dynamicOSObjectCast dynamicOSObjectCastNonARC
#define osObjectCast osObjectCastNonARC
#endif
```

Reply:
IMO, these macros are a hack either way they're defined, and the "NonARC" variants won't realistically go away because we'll always have CF types (always non-ARC) or (in rare cases) use NS types in non-Objective-C[++] sources.

I think a better way to do this is to use `struct ARCEnabled` added by @cdumez in `Source/WTF/wtf/TypeTraits.h` to differentiate `RetainPtr<>` and `OSObjectPtr<>` at build-time by adding a second template parameter.